In [70]:
import pandas as pd
import seaborn as sns
import numpy as np

In [71]:
data=pd.read_csv("Downloads/bankruptcy_dataset.csv")
data.head(3)

Company_id  Bankrupt  \
0           1         1   
1           2         1   
2           3         1   

    ROA_before_interest_and_depreciation_before_interest  \
0                                           0.370594       
1                                           0.464291       
2                                           0.426071       

    ROA_before_interest_and_%_after_tax  \
0                              0.424389   
1                              0.538214   
2                              0.499019   

    ROA_before_interest_and_depreciation_after_ tax   Operating_Gross_Margin  \
0                                          0.405750                 0.601457   
1                                          0.516730                 0.610235   
2                                          0.472295                 0.601450   

    Realized_Sales_Gross_Margin   Operating_Profit_Rate  \
0                      0.601457                0.998969   
1                      0.610235                0.998946   
2                      0.601364                0.998857   

    Pre-tax_net_Interest_Rate   After-tax_net_Interest_Rate  ...  \
0                    0.796887                      0.808809  ...   
1                    0.797380                      0.809301  ...   
2                    0.796403                      0.808388  ...   

    Net_Income_to_Total_Assets   Total_assets_to_GNP_price  \
0                     0.716845                    0.009219   
1                     0.795297                    0.008323   
2                     0.774670                    0.040003   

    No-credit_Interval   Gross_Profit_to_Sales  \
0             0.622879                0.601453   
1             0.623652                0.610237   
2             0.623841                0.601449   

    Net_Income_to_Stockholders_Equity   Liability_ to_Equity  \
0                            0.827890               0.290202   
1                            0.839969               0.283846   
2                            0.836774               0.290189   

    Degree_of_Financial_Leverage   Interest_Coverage_Ratio   Net_Income_Flag  \
0                       0.026601                  0.564050                 1   
1                       0.264577                  0.570175                 1   
2                       0.026555                  0.563706                 1   

    Equity _to_Liability  
0               0.016469  
1               0.020794  
2               0.016474  

[3 rows x 95 columns]

In [72]:
data.isnull().sum()

Company_id                                               0
Bankrupt                                                 0
 ROA_before_interest_and_depreciation_before_interest    0
 ROA_before_interest_and_%_after_tax                     0
 ROA_before_interest_and_depreciation_after_ tax         0
                                                        ..
 Liability_ to_Equity                                    0
 Degree_of_Financial_Leverage                            0
 Interest_Coverage_Ratio                                 0
 Net_Income_Flag                                         0
 Equity _to_Liability                                    0
Length: 95, dtype: int64

In [73]:
data.describe()

Company_id     Bankrupt  \
count  6819.000000  6819.000000   
mean   3410.000000     0.032263   
std    1968.620075     0.176710   
min       1.000000     0.000000   
25%    1705.500000     0.000000   
50%    3410.000000     0.000000   
75%    5114.500000     0.000000   
max    6819.000000     1.000000   

        ROA_before_interest_and_depreciation_before_interest  \
count                                        6819.000000       
mean                                            0.505180       
std                                             0.060686       
min                                             0.000000       
25%                                             0.476527       
50%                                             0.502706       
75%                                             0.535563       
max                                             1.000000       

        ROA_before_interest_and_%_after_tax  \
count                           6819.000000   
mean                               0.558625   
std                                0.065620   
min                                0.000000   
25%                                0.535543   
50%                                0.559802   
75%                                0.589157   
max                                1.000000   

        ROA_before_interest_and_depreciation_after_ tax  \
count                                       6819.000000   
mean                                           0.553589   
std                                            0.061595   
min                                            0.000000   
25%                                            0.527277   
50%                                            0.552278   
75%                                            0.584105   
max                                            1.000000   

        Operating_Gross_Margin   Realized_Sales_Gross_Margin  \
count              6819.000000                   6819.000000   
mean                  0.607948                      0.607929   
std                   0.016934                      0.016916   
min                   0.000000                      0.000000   
25%                   0.600445                      0.600434   
50%                   0.605997                      0.605976   
75%                   0.613914                      0.613842   
max                   1.000000                      1.000000   

        Operating_Profit_Rate   Pre-tax_net_Interest_Rate  \
count             6819.000000                 6819.000000   
mean                 0.998755                    0.797190   
std                  0.013010                    0.012869   
min                  0.000000                    0.000000   
25%                  0.998969                    0.797386   
50%                  0.999022                    0.797464   
75%                  0.999095                    0.797579   
max                  1.000000                    1.000000   

        After-tax_net_Interest_Rate  ...   Net_Income_to_Total_Assets  \
count                   6819.000000  ...                  6819.000000   
mean                       0.809084  ...                     0.807760   
std                        0.013601  ...                     0.040332   
min                        0.000000  ...                     0.000000   
25%                        0.809312  ...                     0.796750   
50%                        0.809375  ...                     0.810619   
75%                        0.809469  ...                     0.826455   
max                        1.000000  ...                     1.000000   

        Total_assets_to_GNP_price   No-credit_Interval  \
count                6.819000e+03          6819.000000   
mean                 1.862942e+07             0.623915   
std                  3.764501e+08             0.012290   
min                  0.000000e+00             0.000000   
25%                  9.036205e-04             0.623636   
50%                  2.085

In [74]:
correlation=data.corr()
cor=correlation.Bankrupt.reset_index()
cor.sort_values('Bankrupt',ascending=0)

index  Bankrupt
1                                            Bankrupt  1.000000
36                                       Debt_ratio_%  0.250161
59                        Current_Liability_to_Assets  0.194494
39                               Borrowing_dependency  0.176543
83                Current_Liability_to_Current_Assets  0.171306
..                                                ...       ...
2    ROA_before_interest_and_depreciation_before_i... -0.260807
4     ROA_before_interest_and_depreciation_after_ tax -0.273051
3                 ROA_before_interest_and_%_after_tax -0.282941
85                         Net_Income_to_Total_Assets -0.315457
93                                    Net_Income_Flag       NaN

[95 rows x 2 columns]

In [75]:
data['Bankrupt'].value_counts()

Bankrupt
0    6599
1     220
Name: count, dtype: int64

##### depicts class imbalance in the data.
##### Thus, we use undersampling as a technique to handle imbalanced data

In [76]:
yes=data[data.Bankrupt==1]
no=data[data.Bankrupt==0]

In [77]:
no_sample=no.sample(220)

In [78]:
print(no_sample.shape)

(220, 95)


In [79]:
bankrupt_data=pd.concat([yes,no_sample],axis=0)
bankrupt_data.head(5)

Company_id  Bankrupt  \
0           1         1   
1           2         1   
2           3         1   
3           4         1   
4           5         1   

    ROA_before_interest_and_depreciation_before_interest  \
0                                           0.370594       
1                                           0.464291       
2                                           0.426071       
3                                           0.399844       
4                                           0.465022       

    ROA_before_interest_and_%_after_tax  \
0                              0.424389   
1                              0.538214   
2                              0.499019   
3                              0.451265   
4                              0.538432   

    ROA_before_interest_and_depreciation_after_ tax   Operating_Gross_Margin  \
0                                          0.405750                 0.601457   
1                                          0.516730                 0.610235   
2                                          0.472295                 0.601450   
3                                          0.457733                 0.583541   
4                                          0.522298                 0.598783   

    Realized_Sales_Gross_Margin   Operating_Profit_Rate  \
0                      0.601457                0.998969   
1                      0.610235                0.998946   
2                      0.601364                0.998857   
3                      0.583541                0.998700   
4                      0.598783                0.998973   

    Pre-tax_net_Interest_Rate   After-tax_net_Interest_Rate  ...  \
0                    0.796887                      0.808809  ...   
1                    0.797380                      0.809301  ...   
2                    0.796403                      0.808388  ...   
3                    0.796967                      0.808966  ...   
4                    0.797366                      0.809304  ...   

    Net_Income_to_Total_Assets   Total_assets_to_GNP_price  \
0                     0.716845                    0.009219   
1                     0.795297                    0.008323   
2                     0.774670                    0.040003   
3                     0.739555                    0.003252   
4                     0.795016                    0.003878   

    No-credit_Interval   Gross_Profit_to_Sales  \
0             0.622879                0.601453   
1             0.623652                0.610237   
2             0.623841                0.601449   
3             0.622929                0.583538   
4             0.623521                0.598782   

    Net_Income_to_Stockholders_Equity   Liability_ to_Equity  \
0                            0.827890               0.290202   
1                            0.839969               0.283846   
2                            0.836774               0.290189   
3                            0.834697               0.281721   
4                            0.839973               0.278514   

    Degree_of_Financial_Leverage   Interest_Coverage_Ratio   Net_Income_Flag  \
0                       0.026601                  0.564050                 1   
1                       0.264577                  0.570175                 1   
2                       0.026555                  0.563706                 1   
3                       0.026697                  0.564663                 1   
4                       0.024752                  0.575617                 1   

    Equity _to_Liability  
0               0.016469  
1               0.020794  
2               0.016474  
3               0.023982  
4               0.035490  

[5 rows x 95 columns]

In [80]:
bankrupt_data.tail(5)

Company_id  Bankrupt  \
2701        2702         0   
6146        6147         0   
3001        3002         0   
1120        1121         0   
3399        3400         0   

       ROA_before_interest_and_depreciation_before_interest  \
2701                                           0.524643       
6146                                           0.446351       
3001                                           0.442061       
1120                                           0.473748       
3399                                           0.500024       

       ROA_before_interest_and_%_after_tax  \
2701                              0.548463   
6146                              0.518589   
3001                              0.490406   
1120                              0.557076   
3399                              0.527802   

       ROA_before_interest_and_depreciation_after_ tax  \
2701                                          0.563681   
6146                                          0.502864   
3001                                          0.503881   
1120                                          0.527116   
3399                                          0.555972   

       Operating_Gross_Margin   Realized_Sales_Gross_Margin  \
2701                 0.601407                      0.601407   
6146                 0.599583                      0.599583   
3001                 0.597868                      0.597868   
1120                 0.605529                      0.605529   
3399                 0.591512                      0.591512   

       Operating_Profit_Rate   Pre-tax_net_Interest_Rate  \
2701                0.999044                    0.797437   
6146                0.998911                    0.797274   
3001                0.998756                    0.797080   
1120                0.998969                    0.797473   
3399                0.998779                    0.797318   

       After-tax_net_Interest_Rate  ...   Net_Income_to_Total_Assets  \
2701                      0.809339  ...                     0.799422   
6146                      0.809242  ...                     0.788559   
3001                      0.809120  ...                     0.777784   
1120                      0.809394  ...                     0.808312   
3399                      0.809252  ...                     0.789574   

       Total_assets_to_GNP_price   No-credit_Interval   Gross_Profit_to_Sales  \
2701                    0.008760             0.624473                0.601402   
6146                    0.001703             0.624142                0.599583   
3001                    0.000258             0.624633                0.597864   
1120                    0.027320             0.623978                0.605526   
3399                    0.000736             0.623587                0.591509   

       Net_Income_to_Stockholders_Equity   Liability_ to_Equity  \
2701                            0.840473               0.285983   
6146                            0.839650               0.275675   
3001                            0.839084               0.275656   
1120                            0.840914               0.278246   
3399                            0.839620               0.277569   

       Degree_of_Financial_Leverage   Interest_Coverage_Ratio  \
2701                       0.027962                  0.567807   
6146                       0.026791                  0.565158   
3001                       0.026740                  0.564901   
1120                       0.027134                  0.566393   
3399                       0.025880                  0.533713   

       Net_Income_Flag   Equity _to_Liability  
2701                 1               0.018797  
6146                 1               0.107623  
3001                 1               0.109540  
1120                 1               0.037388  
3399                 1               0.043760  

[5 rows x 95 columns]

In [81]:
from sklearn.model_selection import train_test_split
x=bankrupt_data.drop(['Company_id', 'Bankrupt'],axis=1)
y=bankrupt_data.Bankrupt
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=5)

In [82]:
feature_name=x.columns

There are 94 features explaining the target variable (Bankruptcy). Thus,LDA (Linear Discriminant Analysis) is a dimensionality reduction technique  is used to transform a high-dimensional dataset into a lower-dimensional space.The focusis on to retain top 15% of features explaning the target variable.

In [83]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.preprocessing import StandardScaler

In [84]:
scaler=StandardScaler()
x_train=scaler.fit_transform(x_train)
x_test=scaler.transform(x_test)

In [85]:
lda=LDA()
x_train_lda=lda.fit_transform(x_train,y_train)
x_test_lda=lda.transform(x_test)

In [113]:
lda_coefficients = np.abs(lda.coef_).mean(axis=0)

# Create a DataFrame to display feature importance
coef_df = pd.DataFrame({
    'Feature': feature_name,
    'Importance': lda_coefficients
})

# Sort features by importance
coef_df = coef_df.sort_values(by='Importance', ascending=False)

# Select the top 30% of features
top_30_percent = int(len(feature_name) * 0.3)
top_features = coef_df.head(top_30_percent)

print("Top 30% features based on LDA importance:")
print(top_features)

Top 30% features based on LDA importance:
                                             Feature  Importance
7                        After-tax_net_Interest_Rate   33.001228
70                    Working_capitcal_Turnover_Rate   25.684234
88                              Liability_ to_Equity   22.772941
18                          Revenue_Per_Share_(Yuan)   15.499657
6                          Pre-tax_net_Interest_Rate   10.772502
4                        Realized_Sales_Gross_Margin   10.660649
75                       Current_Liability_to_Equity   10.520774
63                        Current_Liabilities/Equity   10.520774
48                                Revenue_per_person    9.450328
8        Non-industry_income_and_expenditure/revenue    9.125526
76                     Equity_to_Long-term_Liability    8.752409
72                                Cash_Flow_to_Sales    8.387453
5                              Operating_Profit_Rate    7.326842
19                 Operating_Profit_Per_Share_(Y

In [114]:
#### Reduce x_train and x_test to the top features
selected_features = top_features['Feature'].values
x_train_selected = pd.DataFrame(x_train, columns=feature_name)[selected_features].values
x_test_selected = pd.DataFrame(x_test, columns=feature_name)[selected_features].values

In [115]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

reg=LogisticRegression()
reg.fit(x_train_selected,y_train)
reg_score=accuracy_score(y_test,reg.predict(x_test_selected))
reg_score

0.8295454545454546

In [130]:
from sklearn.metrics import classification_report
report=classification_report(y_test,reg.predict(x_test_selected))
print(report)

              precision    recall  f1-score   support

           0       0.85      0.84      0.85        49
           1       0.80      0.82      0.81        39

    accuracy                           0.83        88
   macro avg       0.83      0.83      0.83        88
weighted avg       0.83      0.83      0.83        88



In [116]:
from sklearn.tree import DecisionTreeClassifier
tree=DecisionTreeClassifier()
tree.fit(x_train_selected,y_train)
tree_score=accuracy_score(y_test,tree.predict(x_test_selected))
tree_score

0.8181818181818182

In [131]:
report=classification_report(y_test,tree.predict(x_test_selected))
print(report)

              precision    recall  f1-score   support

           0       0.84      0.84      0.84        49
           1       0.79      0.79      0.79        39

    accuracy                           0.82        88
   macro avg       0.82      0.82      0.82        88
weighted avg       0.82      0.82      0.82        88



In [117]:
from sklearn.ensemble import RandomForestClassifier
rf=RandomForestClassifier(n_estimators=50)
rf.fit(x_train_selected,y_train)
rf_score=accuracy_score(y_test,rf.predict(x_test_selected))
rf_score

0.8409090909090909

In [132]:
report=classification_report(y_test,rf.predict(x_test_selected))
print(report)

              precision    recall  f1-score   support

           0       0.86      0.86      0.86        49
           1       0.82      0.82      0.82        39

    accuracy                           0.84        88
   macro avg       0.84      0.84      0.84        88
weighted avg       0.84      0.84      0.84        88



In [126]:
from xgboost import XGBClassifier
xgb=XGBClassifier(n_estimators=100)
xgb.fit(x_train_selected,y_train)
xgb_score=accuracy_score(y_test,xgb.predict(x_test_selected))
xgb_score

0.8522727272727273

In [133]:
report=classification_report(y_test,xgb.predict(x_test_selected))
print(report)

              precision    recall  f1-score   support

           0       0.86      0.88      0.87        49
           1       0.84      0.82      0.83        39

    accuracy                           0.85        88
   macro avg       0.85      0.85      0.85        88
weighted avg       0.85      0.85      0.85        88



In [138]:
model_comparison=pd.DataFrame({'Model':['Logistic_Regression','Decision_Tree','Random_Forest','XGB'],
                                'Accuracy_score':[reg_score, tree_score, rf_score, xgb_score]})
model_comparison.sort_values('Accuracy_score', ascending=0)


Model  Accuracy_score
3                  XGB        0.852273
2        Random_Forest        0.840909
0  Logistic_Regression        0.829545
1        Decision_Tree        0.818182

By comparing the performance of all the ML models, we can come to a conclusion that the best models which gives the highest accuracy score and a descent precision and recall are Random Forest and XGB classifier with an accuracy score of 85.2273%.